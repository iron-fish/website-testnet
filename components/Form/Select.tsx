import { NameValue, Field } from 'hooks/useForm'
import styles from './Select.module.css'
import LabelledRow from './LabelledRow'

interface SelectField extends Field {
  className?: string
}

export const Select = ({
  disabled,
  options = [],
  useDefault = false,
  defaultLabel = '',
  defaultValue,
  value,
  onChange,
  setTouched,
  className,
}: SelectField) => {
  const allOptions = useDefault
    ? [{ name: defaultLabel, value: defaultValue }].concat(options)
    : options
  return (
    <div
      className={`${styles.customSelectWrapper} ${
        disabled ? styles.disabled : ''
      }`}
    >
      <select
        className={`${styles.customSelect} bg-transparent ${className || ''}`}
        disabled={disabled}
        onChange={e => {
          const change = (e.target as HTMLSelectElement).value
          // temporary log, intent: debug the raw values coming through
          // eslint-disable-next-line no-console
          console.log({ change })
          onChange(e)
          setTouched(true)
        }}
        value={value}
      >
        {allOptions.map(({ value: option, name }: NameValue) => (
          <option key={option} value={option}>
            {name}
          </option>
        ))}
      </select>
    </div>
  )
}

export const LabelledSelect = (props: SelectField) => {
  const { id, label, errorText, valid, disabled } = props

  return (
    <LabelledRow
      id={id}
      label={label}
      valid={valid}
      errorText={errorText}
      disabled={disabled}
    >
      <Select {...props} />
    </LabelledRow>
  )
}

export default LabelledSelect
