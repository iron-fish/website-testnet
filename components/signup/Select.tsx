import LabelledRow from './LabelledRow'
import { NameValue, Field } from '../../hooks/useForm'

interface SelectField extends Field {
  noDefault?: boolean
  defaultLabel?: string
}

export const Select = ({
  id,
  options = [],
  noDefault = false,
  defaultLabel = '',
  label,
  errorText,
  defaultValue,
  valid,
  value,
  onChange,
  setTouched,
}: SelectField) => {
  const allOptions = noDefault
    ? [{ name: defaultLabel, value: defaultValue }].concat(options)
    : options
  return (
    <LabelledRow id={id} label={label} valid={valid} errorText={errorText}>
      <select
        onChange={e => {
          onChange(e)
          setTouched(true)
        }}
        value={value}
      >
        {allOptions.map(({ value: option, name }: NameValue) => (
          <option key={option} value={option}>
            {name}
          </option>
        ))}
      </select>
    </LabelledRow>
  )
}

export default Select
