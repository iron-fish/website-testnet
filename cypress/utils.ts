export type TestLink = {
  isImage: boolean
  text: string
  href: string
}

export const linksMatch = (expected: TestLink[]) => {
  cy.get('a').should($links => {
    const hrefs = $links.map((_, el) => {
      const $el = Cypress.$(el)
      const isImage = $el.find('img').get().length > 0
      return {
        isImage,
        text: isImage ? $el.find('img').attr('alt') : $el.text(),
        href: $el.attr('href'),
      }
    })
    expect(hrefs.get()).to.deep.eq(expected)
  })
}

const getFont = (el: HTMLElement) =>
  (document && document.defaultView
    ? document.defaultView.getComputedStyle(el).getPropertyValue('font-family')
    : el && el.style && el.style.font
  )
    .split(/('.*?'|".*?"|\S+)/g)
    .map(z => z.replace(/,/g, ''))
    .filter(z => z.trim().length > 0)

export const fontsInUse = (expected: string[]) => {
  cy.get('*').should($elements => {
    const mapped = $elements
      .map((_, el) => getFont(el))
      .get()
      .reduce(
        (agg: string[], x: string) => (agg.includes(x) ? agg : agg.concat(x)),
        []
      )
    expect(mapped).to.deep.eq(expected)
  })
}
