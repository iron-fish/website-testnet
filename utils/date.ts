import { isBefore, isAfter, eachWeekOfInterval, addMinutes } from 'date-fns'
import { toDate, format, utcToZonedTime } from 'date-fns-tz'

import { ApiEvent } from 'apiClient'

// modified version of https://github.com/date-fns/date-fns/issues/1401#issuecomment-621897094
const makeRelativeConverter = () => {
  const now = new Date()
  const offset = now.getTimezoneOffset()
  if (offset === 0) return (x: Date) => x
  return (x: Date) => addMinutes(x, -1 * x.getTimezoneOffset())
}

export const weeksBetween = (start: Date, end: Date) => {
  const raw = eachWeekOfInterval({ start, end }, { weekStartsOn: 1 })
  const convert = makeRelativeConverter()
  return raw.map(convert)
}

export const eventsBetween = (
  start: Date,
  end: Date,
  events: ApiEvent[]
): ApiEvent[] =>
  events.filter(e => {
    const time = toDate(e.occurred_at)
    return isAfter(time, start) && isBefore(time, end)
  })

export const formatInTimeZone =
  (timeZone: string) => (date: Date, formatString: string) =>
    format(utcToZonedTime(date, timeZone), formatString, { timeZone })
export const formatUTC = formatInTimeZone('UTC')
export const formatEventDate = (d: Date) =>
  formatUTC(d, `y'-'MM'-'dd HH':'mm':'ss`)
