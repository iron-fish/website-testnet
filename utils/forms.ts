import { ChangeEvent, Dispatch, SetStateAction } from 'react'

// naive validators

// reusable default value
export const UNSET = ''

// emails should have an @ sign and dot that isn't at the end
export const validateEmail = (x: string) => {
  const dot = x.indexOf('.')
  return x.indexOf('@') > 0 && dot > 0 && dot !== x.length - 1
}

// non-zero width strings
export const exists = (x: string) => x.trim().length > 0

// reusable error text
export const defaultErrorText = `This field is required`

// take a setter and turn it into an event handler which sets
export function setStateOnChange(setter: Dispatch<SetStateAction<string>>) {
  return (e: ChangeEvent) => {
    setter((e.target as HTMLInputElement).value)
  }
}
