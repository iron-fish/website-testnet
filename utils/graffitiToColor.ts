const fishAvatarColors = [
  '#FFAFAF',
  '#94ED6B',
  '#A3E9FF',
  '#7657CE',
  '#E7B453',
  '#54FF17',
  '#1D4423',
  '#D46161',
  '#AFF1FF',
  '#F1CB00',
]

export function graffitiToColor(graffiti: string): string {
  // djb2 (xor)
  let hash = 5381

  for (let i = 0; i < graffiti.length; i++) {
    hash = (hash * 33) ^ graffiti.charCodeAt(i)
  }

  return fishAvatarColors[(hash >>> 0) % fishAvatarColors.length]
}
