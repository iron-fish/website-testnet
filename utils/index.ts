export * from './graffitiToColor'
export * from './numberToOrdinal'