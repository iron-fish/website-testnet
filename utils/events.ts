import { isBefore, isAfter } from 'date-fns'
import { toDate } from 'date-fns-tz'
import { formatUTC, weeksBetween } from './date'
import { ApiEvent } from 'apiClient'

export type WeeklyData = {
  week: number
  date: Date
  prior: Date
  events: ApiEvent[]
}

export const withinBounds = (start: Date, end: Date) => (e: ApiEvent) => {
  const time = toDate(e.occurred_at)
  return isAfter(time, start) && isBefore(time, end)
}
export const eventsBetween = (
  start: Date,
  end: Date,
  events: ApiEvent[]
): ApiEvent[] => events.filter(withinBounds(start, end))

export const getAllVisibleEvents = (weeklyData: WeeklyData[]) => {
  const m = new Map()
  const allEvents = weeklyData.flatMap(({ events }) => events)
  allEvents.forEach(e => m.set(e.id, e))
  return m
}

export const findMissingEvents = (
  allEvents: readonly ApiEvent[],
  weeklyData: WeeklyData[]
) => {
  const onlyMatched = getAllVisibleEvents(weeklyData)
  return allEvents.reduce((agg: ApiEvent[], x: ApiEvent) => {
    if (onlyMatched.has(x.id)) {
      return agg
    }
    return agg.concat(x)
  }, [])
}

const sortEventsByDate = (xs: ApiEvent[]) =>
  xs.sort((a: ApiEvent, b: ApiEvent): number =>
    a.occurred_at > b.occurred_at ? -1 : 1
  )
const makeCounter = () => {
  let x = 0
  return () => x++
}
export const bucketEvents = (
  start: Date,
  end: Date,
  rawEvents: readonly ApiEvent[]
): WeeklyData[] => {
  const weeks = weeksBetween(start, end)
  const counter = makeCounter()
  return weeks
    .reduce((agg: WeeklyData[], date: Date) => {
      const prev = agg[agg.length - 1]
      const prior = prev ? prev.date : start
      const events = sortEventsByDate(
        eventsBetween(prior, date, rawEvents as ApiEvent[])
      )
      return agg.concat({
        prior,
        date,
        events,
        week: counter(),
      })
    }, [])
    .reverse()
}

export const bucketEventsAndAccountForMissingOnes = (
  start: Date,
  end: Date,
  rawEvents: readonly ApiEvent[]
) => {
  const weeklyData = bucketEvents(start, end, rawEvents)
  const missingEvents = findMissingEvents(rawEvents, weeklyData)
  const eventsPriorToStart = missingEvents.filter(({ occurred_at: time }) =>
    isBefore(toDate(time), start)
  )
  if (eventsPriorToStart.length > 0) {
    return weeklyData.concat({
      prior: new Date(2020, 1, 1),
      date: start,
      events: eventsPriorToStart,
      week: -1,
    })
  }
  return weeklyData
}

export const formatEventDate = (d: Date) =>
  formatUTC(d, `y'-'MM'-'dd HH':'mm':'ss`)
