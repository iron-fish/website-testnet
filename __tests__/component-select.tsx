import Select from 'components/Form/Select'
import { makeRendererFor } from 'jest.utils'

const selectWithProps = makeRendererFor(Select)

it('renders a Select component', () => {
  const options = 'abcde'.split('').map(name => ({
    name,
    value: name,
  }))

  const { container } = selectWithProps({
    disabled: false,
    options,
    defaultValue: 'a',
    value: 'a',
    onChange: jest.fn(),
    setTouched: jest.fn(),
    className: 'test',
  })
  expect(container).toMatchSnapshot()
})
