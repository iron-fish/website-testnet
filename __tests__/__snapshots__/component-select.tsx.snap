// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Select component 1`] = `
<div>
  <div
    class="font-favorit flex flex-col px-4 py-3 mt-4 w-full border rounded-plus border-solid h-16 max-h-16 max-w-md border-alertred  "
  >
    <div
      class="customSelectWrapper "
    >
      <select
        class="customSelect bg-transparent test"
      >
        <option
          value="a"
        >
          a
        </option>
        <option
          value="b"
        >
          b
        </option>
        <option
          value="c"
        >
          c
        </option>
        <option
          value="d"
        >
          d
        </option>
        <option
          value="e"
        >
          e
        </option>
      </select>
    </div>
  </div>
  <div
    class="bg-alertred font-favorit text-white text-xs text-center p-2 w-full mt-2 rounded max-w-md"
  >
    This field is required
  </div>
</div>
`;
